## Human single-cell RNA-seq using Seurat standard analysis workflow
This is part of a 2-day training course -- *Human Gene Expression and Omic Data Analysis Course* -- designed to encourage PhDs and Postdocs to start their own bioinformatic analysis journey. The demo environment was set up via [Posit Cloud](https://posit.cloud/plans/instructor) which provided hassle-free dependency management.
We will use the 19 human PBMC (peripheral blood mononuclear cells) data sets with high and low HBsAG (Hepatitis B surface antigen) levels from the NUC (Nucleos(t)ide analogues i.e., tenofovir or entecavir) treated chronic HBV (hepatitis B virus) patients, from the following paper.
> Beudeker, B. J. B., Osmani, Z., van Oord, G. W., Groothuismink, Z. M. A., de Knegt, R. J., Hoogenboezem, R. M., Bindels, E. M. J., van de Werken, H. J. G., & Boonstra, A. (2023). Association of HBsAg levels with differential gene expression in NK, CD8 T, and memory B cells in treated patients with chronic HBV. JHEP reports : innovation in hepatology, 6(2), 100980. https://doi.org/10.1016/j.jhepr.2023.100980
> 
We will load the count matrices, cell barcodes, and gene names (i.e. [the output of Cell Ranger pipeline](https://www.10xgenomics.com/support/software/cell-ranger/latest/analysis/outputs/cr-outputs-mex-matrices)) downloaded from [GEO accession number GSE247322](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE247322). [Seurat](https://satijalab.org/seurat/) is an R package designed for QC, analysis, and exploration of single-cell RNA-seq data. Seurat aims to enable users to identify and interpret sources of heterogeneity from single-cell transcriptomic measurements, and to integrate diverse types of single-cell data. 
